pylate/server/server.py

```diff
@@ -0,0 +1,99 @@
+import argparse
+from typing import List
+
+import batched
+import uvicorn
+from fastapi import FastAPI, HTTPException
+from pydantic import BaseModel
+
+from pylate import models
+
+app = FastAPI()
+
+
+class EmbeddingRequest(BaseModel):
+    input: List[str] | str
+    is_query: bool = True
+    model: str = "lightonai/colbertv2.0"
+
+
+class EmbeddingResponse(BaseModel):
+    data: List[dict]
+    model: str
+    usage: dict
+
+
+def wrap_encode_function(model, **kwargs):
+    def wrapped_encode(sentences):
+        return model.encode(sentences, **kwargs)
+
+    return wrapped_encode
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Run FastAPI ColBERT serving server with specified host, port, and model."
+    )
+    parser.add_argument(
+        "--host", type=str, default="0.0.0.0", help="Host to run the server on"
+    )
+    parser.add_argument(
+        "--port", type=int, default=8002, help="Port to run the server on"
+    )
+    parser.add_argument(
+        "--model",
+        type=str,
+        default="lightonai/colbertv2.0",
+        help="Model to serve, can be an HF model or a path to a model",
+    )
+    return parser.parse_args()
+
+
+args = parse_args()
+
+# We need to load the model here so it is shared for every request
+model = models.ColBERT(args.model)
```

Comment:
We should be loading the model at startup only ?

Reply:
On this specific point, I think it would be better to create a class 

```python
class Model:
    """""""
    def __init__(self) ->:
        """""
        self.model = None

    def encode(self, text) ->:
         """""
         return self.model.encode(text)

    def load(self, model) ->:
        self.model = models.ColBERT(args.model)
```

Initialize the model just like you initialize the ColBERT model and in the dedicated endpoint which is launched at startup by fastapi execute `model.load(model)` otherwise you will load the model at each call no ?

Reply:
useful resource: https://fastapi.tiangolo.com/advanced/events/#lifespan

Reply:
Closing this as the model is only loaded once